### PR TITLE
Fix warnings and tests for 32-bit ARM

### PR DESCRIPTION
**Pull request by emersonknapp**
Fixes https://github.com/eProsima/Fast-CDR/issues/52

This is an attempt to fix the warnings and test failures in armhf builds.

This ROS2 build shows the warnings that were being raised https://ci.ros2.org/view/All/job/test_ci_linux-armhf/13/warnings23Result/
* Fix `useless-cast` warnings by removing unnecessary use of `uint32_t` when working with `size_t` values
* Fix `cast-align` warnings by doing `memcpy` instead of `reinterpret_cast`s
* Fix test suite by changing `_WIN32` preprocessor guards to also guard ARM32 builds

Please let me know if you would like this split out into separate changes, or if you would prefer to keep it as-is for a single review.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/53/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/57/)

**Comment by emersonknapp** [comment]
I am able reproduce the x86_64 failures locally, I'll take another pass at this. Any early feedback would still be greatly appreciated.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/54/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/58/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/55/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/59/)

File: include/fastcdr/config.h.in [hunk 1]
```diff
@@ -46,4 +46,12 @@
 #define __BIG_ENDIAN__ @__BIG_ENDIAN__@
 #endif
 
+#if defined(__ARM_ARCH) && __ARM_ARCH <= 7
+#define FASTCDR_32BIT
+#endif
+
+#if defined(_WIN32)
```
Comment (LuisGP) [inline include/fastcdr/config.h.in]: I think this isn't enough, `_WIN32` is defined in Win64 too.

```suggestion
#if defined(_WIN32) and not defined(_WIN64)
```

I think the suggestion ensures we are in a pure Win32 platform.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/56/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/60/)

File: include/fastcdr/config.h.in (hunk 1 above)
Comment (emersonknapp) [reply include/fastcdr/config.h.in]: I just committed this change, but in this PR I switched over all cases of `if defined(_WIN32)`. Those seemed intentional, and now would not be hit by WIN64 builds.

The build with this stricter check makes the Windows build unstable - it looks like some of the `#if defined(_WIN32)` blocks were not specific to 32bit builds.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/57/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/61/)

**Comment by emersonknapp** [comment]
Hi @LuisGP or @richiware - I am having a hard time reproducing the Windows build - I am able to build on Windows but not using the same arguments as your Jenkins server. Could you share the eprosima/windows-ci docker container so that I can reproduce the build exactly?

File: include/fastcdr/config.h.in (hunk 1 above)
Comment (LuisGP) [reply include/fastcdr/config.h.in]: What do you mean? `and` and `not` are *alternative operators* and should be supported by all C++11 compliant compilators.
I personally prefer using them only in the preprocessor, but it's a matter of style. In any case, shouldn't be incorrect. Please, correct me if I'm wrong.

Comment (emersonknapp) [reply include/fastcdr/config.h.in]: Hmm... maybe there was a non-ascii character in there then. The Windows builds complained specifically about that line. My mistake

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/58/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/62/)

File: include/fastcdr/config.h.in (hunk 1 above)
Comment (emersonknapp) [reply include/fastcdr/config.h.in]: I just tested this explicitly with a commit. GCC accepts the alternative operators, but the Windows build preprocessor does not.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/59/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/63/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Linux/60/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastCDR%20Manual%20Windows/64/)

File: include/fastcdr/config.h.in (hunk 1 above)
Comment (LuisGP) [reply include/fastcdr/config.h.in]: We did some testing here too. Seems that Visual Studio needs to set a flag in the compiler options to allow them.
It is OK to use traditional operators to simplify the overall process.
Thank you.

**Comment by emersonknapp** [comment]
Thanks @luisgp for confirming that. Could somebody on your team review this code? It fixes the arm32 issues and doesn't change any of the existing builds.

**Pull request merged**